File: samples/java/SupportChat/Server/Support/src/main/java/systems/zlink/samples/supportchat/server/support/spots/conversationspot/ConversationSpot.java
package systems.zlink.samples.supportchat.server.support.spots.conversationspot;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import systems.zlink.framework.messaging.ZLinkMessage;
import systems.zlink.framework.spots.ZLinkSpot;
import systems.zlink.framework.spots.ZLinkSpotActorJoinResult;
import systems.zlink.framework.spots.ZLinkSpotContext;
import systems.zlink.framework.spots.ZLinkSpotCreateResponse;
import systems.zlink.framework.spots.ZLinkTimer;
import systems.zlink.samples.supportchat.server.configuration.SampleNames;
import systems.zlink.samples.supportchat.server.configuration.SampleTimings;
import systems.zlink.samples.supportchat.server.support.actors.SupportActorDirectory;
import systems.zlink.samples.supportchat.server.support.actors.SupportUserActor;
import systems.zlink.samples.supportchat.server.support.application.AgentAssignmentService;
import systems.zlink.samples.supportchat.server.support.domain.Conversation;
import systems.zlink.samples.supportchat.server.support.infrastructure.ConversationContracts;
import systems.zlink.samples.supportchat.server.support.infrastructure.ConversationNotificationPublisher;
import systems.zlink.samples.supportchat.shared.contracts.Messages;

import java.time.Duration;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public final class ConversationSpot implements ZLinkSpot<SupportUserActor> {
    private static final Logger logger = LoggerFactory.getLogger(ConversationSpot.class);
    private final ZLinkSpotContext context;
    private final AgentAssignmentService assignment;
    private final SupportActorDirectory directory;
    private final ConversationNotificationPublisher notifications;
    private final Map<String, SupportUserActor> actors = new LinkedHashMap<>();
    private final Set<String> pendingJoins = new HashSet<>();
    private Conversation conversation;
    private ZLinkTimer idleTimer;

    public ConversationSpot(
            ZLinkSpotContext context,
            AgentAssignmentService assignment,
            SupportActorDirectory directory,
            ConversationNotificationPublisher notifications) {
        this.context = context;
        this.assignment = assignment;
        this.directory = directory;
        this.notifications = notifications;
    }

    @Override
    public ZLinkSpotContext context() {
        return context;
    }

    @Override
    public CompletionStage<ZLinkSpotCreateResponse> onCreate(ZLinkMessage request) {
        ConversationCreateReq create = request.decode(ConversationCreateReq.class);
        conversation =
                new Conversation(
                        context.spotId(),
                        create.subject(),
                        create.customerActorId(),
                        create.customerDisplayName(),
                        create.createdAtUnixMs(),
                        new Conversation.Policy(
                                SampleTimings.IdleTimeout, SampleTimings.CloseGraceTimeout, 500));
        logger.info("supportchat-conversation created conversation={}", context.spotId());
        logger.info(
                "supportchat-conversation status={} conversation={}",
                SampleNames.Statuses.WaitingForAgent,
                context.spotId());
        return CompletableFuture.completedFuture(ZLinkSpotCreateResponse.accept());
    }

    @Override
    public CompletionStage<Void> onInitialize() {
        return context.addTimer(
                        "conversation-idle",
                        Duration.ofMillis(200),
                        ConversationIdleTimerHandler.class,
                        null)
                .thenAccept(timer -> idleTimer = timer);
    }

    @Override
    public CompletionStage<Void> onClosing() {
        return idleTimer == null ? CompletableFuture.completedFuture(null) : idleTimer.cancel();
    }

    @Override
    public CompletionStage<ZLinkSpotActorJoinResult> onActorJoin(
            String actorId, ZLinkMessage request) {
        request.decode(Messages.JoinConversationReq.class);
        pendingJoins.add(actorId);
        return CompletableFuture.completedFuture(
                ZLinkSpotActorJoinResult.accept(
                        new Messages.JoinConversationRes(
                                false,
                                actorId,
                                ConversationContracts.state(requireConversation().snapshot()))));
    }

    @Override
    public CompletionStage<Void> onJoinedActor(SupportUserActor actor) {
        if (!pendingJoins.remove(actor.actorId())) {
            return CompletableFuture.failedFuture(
                    new IllegalStateException("joined actor does not have pending admission"));
        }
        directory.remember(actor);
        if (SampleNames.Roles.Agent.equals(actor.role())) {
            actors.put(actor.participantId(), actor);
            publish(
                    requireConversation()
                            .joinAgent(
                                    actor.participantId(),
                                    actor.displayName(),
                                    System.currentTimeMillis()));
            logger.info(
                    "supportchat-conversation agent-joined conversation={} agent={}",
                    context.spotId(),
                    actor.participantId());
        } else {
            actor.joinConversation(requireConversation().snapshot().conversationId());
            actors.put(actor.participantId(), actor);
            assignAgent();
        }
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onLeaveActor(SupportUserActor actor) {
        pendingJoins.remove(actor.actorId());
        actors.remove(actor.participantId());
        return CompletableFuture.completedFuture(null);
    }

    public Messages.JoinConversationRes refreshMembership(SupportUserActor actor) {
        actors.put(actor.participantId(), actor);
        return new Messages.JoinConversationRes(
                false,
                actor.actorId(),
                ConversationContracts.state(requireConversation().snapshot()));
    }

    public Messages.SendChatMessageRes sendMessage(
            SupportUserActor actor, Messages.SendChatMessageReq request) {
        Conversation.Change change =
                requireConversation()
                        .sendMessage(
                                actor.participantId(), request.text(), System.currentTimeMillis());
        publish(change);
        Conversation.Message message =
                change.events().stream()
                        .map(Conversation.Event::message)
                        .filter(Objects::nonNull)
                        .findFirst()
                        .orElseThrow();
        return new Messages.SendChatMessageRes(
                ConversationContracts.message(message),
                ConversationContracts.state(change.state()));
    }

    public void setTyping(SupportUserActor actor, Messages.SetTypingMsg message) {
        publish(requireConversation().setTyping(actor.participantId(), message.isTyping()));
    }

    public Messages.CloseConversationRes close(
            SupportUserActor actor, Messages.CloseConversationReq request) {
        Conversation.Change change =
                requireConversation().close(actor.participantId(), request.reason());
        publish(change);
        return new Messages.CloseConversationRes(ConversationContracts.state(change.state()));
    }

    public void checkIdle() {
        publish(requireConversation().markIdle(System.currentTimeMillis()));
    }

    // --8<-- [start:doc-sc-assign]
    private void assignAgent() {
        AgentAssignmentService.AvailableAgent assigned =
                assignment.assignForConversation(requireConversation().snapshot().conversationId());
        if (assigned == null) {
            return;
        }
        notifications.assigned(assigned, requireConversation().snapshot(), directory);
        logger.info(
                "support conversation: assigned. conversation={}, roster={}",
                context.spotId(),
                assigned.rosterActorId());
    }

    // --8<-- [end:doc-sc-assign]

    private void publish(Conversation.Change change) {
        notifications.publish(change, actors, assignment);
        for (Conversation.Event event : change.events()) {
            Messages.ConversationState state = ConversationContracts.state(event.state());
            logger.info(
                    "supportchat-conversation status={} conversation={}",
                    state.status(),
                    state.conversationId());
        }
    }

    private Conversation requireConversation() {
        if (conversation == null) {
            throw new IllegalStateException("Conversation has not been created");
        }
        return conversation;
    }
}

File: samples/kotlin/SupportChat/Server/Support/src/main/kotlin/systems/zlink/samples/kotlin/supportchat/server/support/infrastructure/zlink/spots/conversationspot/ConversationSpot.kt
package systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.spots.conversationspot

import java.time.Duration
import kotlinx.coroutines.future.await
import org.slf4j.LoggerFactory
import systems.zlink.framework.kotlin.ZLinkSuspendingSpot
import systems.zlink.framework.messaging.ZLinkMessage
import systems.zlink.framework.spots.ZLinkSpotActorJoinResult
import systems.zlink.framework.spots.ZLinkSpotClosingContext
import systems.zlink.framework.spots.ZLinkSpotContext
import systems.zlink.framework.spots.ZLinkSpotCreateResponse
import systems.zlink.framework.spots.ZLinkTimer
import systems.zlink.samples.kotlin.supportchat.server.configuration.ConversationStatuses
import systems.zlink.samples.kotlin.supportchat.server.configuration.SampleTimings
import systems.zlink.samples.kotlin.supportchat.server.configuration.SupportChatRoles
import systems.zlink.samples.kotlin.supportchat.server.support.application.AgentAssignmentService
import systems.zlink.samples.kotlin.supportchat.server.support.domain.Conversation
import systems.zlink.samples.kotlin.supportchat.server.support.domain.ConversationChange
import systems.zlink.samples.kotlin.supportchat.server.support.domain.ConversationEventKind
import systems.zlink.samples.kotlin.supportchat.server.support.domain.ConversationPolicy
import systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.ConversationContracts
import systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.actors.SupportActorDirectory
import systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.actors.SupportUserActor
import systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.spots.conversationspot.handlers.ConversationIdleTimerHandler
import systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.spots.conversationspot.notifications.ConversationNotificationPublisher
import systems.zlink.samples.kotlin.supportchat.shared.contracts.CloseConversationReq
import systems.zlink.samples.kotlin.supportchat.shared.contracts.CloseConversationRes
import systems.zlink.samples.kotlin.supportchat.shared.contracts.JoinConversationReq
import systems.zlink.samples.kotlin.supportchat.shared.contracts.JoinConversationRes
import systems.zlink.samples.kotlin.supportchat.shared.contracts.SendChatMessageReq
import systems.zlink.samples.kotlin.supportchat.shared.contracts.SendChatMessageRes
import systems.zlink.samples.kotlin.supportchat.shared.contracts.SetTypingMsg

class ConversationSpot(
    override val context: ZLinkSpotContext,
    private val notifications: ConversationNotificationPublisher,
    private val assignment: AgentAssignmentService,
    private val directory: SupportActorDirectory,
) : ZLinkSuspendingSpot<SupportUserActor>() {
    private val actors = linkedMapOf<String, SupportUserActor>()
    private val pendingJoins = mutableSetOf<String>()
    private var conversation: Conversation? = null
    private var idleTimer: ZLinkTimer? = null

    override suspend fun onCreateSuspending(request: ZLinkMessage): ZLinkSpotCreateResponse {
        val create = request.decode(ConversationCreateReq::class.java)
        val conversationId = context.spotId()
        conversation =
            Conversation(
                conversationId = conversationId,
                subject = create.subject,
                customerActorId = create.customerActorId,
                customerDisplayName = create.customerDisplayName,
                createdAtUnixMs = create.createdAtUnixMs,
                policy =
                    ConversationPolicy(
                        SampleTimings.IdleTimeout,
                        SampleTimings.CloseGraceTimeout,
                        500,
                    ),
            )
        logger.info("supportchat-conversation created conversation={}", conversationId)
        logger.info(
            "supportchat-conversation status={} conversation={}",
            ConversationStatuses.WaitingForAgent,
            conversationId,
        )
        return ZLinkSpotCreateResponse.accept()
    }

    override suspend fun onInitializeSuspending() {
        idleTimer =
            context
                .addTimer(
                    "conversation-idle",
                    Duration.ofMillis(200),
                    ConversationIdleTimerHandler::class.java,
                    null,
                )
                .await()
    }

    override suspend fun onClosingSuspending(context: ZLinkSpotClosingContext) {
        idleTimer?.cancel()?.await()
    }

    override suspend fun onActorJoinSuspending(
        actorId: String,
        request: ZLinkMessage,
    ): ZLinkSpotActorJoinResult {
        val conversation = requireConversation()
        request.decode(JoinConversationReq::class.java)
        pendingJoins += actorId
        return ZLinkSpotActorJoinResult.accept(
            JoinConversationRes(
                scheduled = false,
                state = ConversationContracts.toState(conversation.snapshot()),
            )
        )
    }

    override suspend fun onJoinedActorSuspending(actor: SupportUserActor) {
        check(pendingJoins.remove(actor.actorId)) {
            "joined actor does not have a pending admission"
        }
        val conversation = requireConversation()
        if (actor.role == SupportChatRoles.Agent) {
            val change = joinAgent(actor)
            publishChange(change)
            logger.info(
                "supportchat-conversation agent-joined conversation={} agent={}",
                conversation.conversationId,
                actor.participantId,
            )
            return
        }

        actor.joinConversation(conversation.conversationId)
        actors[actor.participantId] = actor
        assignAgent()
        logger.info(
            "support conversation: actor joined. conversation={}, participant={}, role={}",
            conversation.conversationId,
            actor.participantId,
            actor.role,
        )
    }

    override suspend fun onLeaveActorSuspending(actor: SupportUserActor) {
        pendingJoins.remove(actor.actorId)
        actors.remove(actor.participantId)
    }

    // --8<-- [start:doc-sc-idle-timer]
    suspend fun checkIdle() {
        val change = conversation?.markIdle(System.currentTimeMillis()) ?: return
        publishChange(change)
    }

    // --8<-- [end:doc-sc-idle-timer]

    fun refreshMembership(actor: SupportUserActor): JoinConversationRes {
        val conversation = requireConversation()
        actors[actor.participantId] = actor
        logger.info(
            "support conversation: membership refreshed. conversation={}, participant={}",
            conversation.conversationId,
            actor.participantId,
        )
        return JoinConversationRes(
            scheduled = false,
            state = ConversationContracts.toState(conversation.snapshot()),
        )
    }

    suspend fun sendMessage(
        actor: SupportUserActor,
        request: SendChatMessageReq,
    ): SendChatMessageRes {
        val change =
            requireConversation()
                .sendMessage(actor.participantId, request.text, System.currentTimeMillis())
        publishChange(change)
        val message =
            change.events.single { it.kind == ConversationEventKind.MessageAppended }.message
                ?: error("Message event was not created.")
        return SendChatMessageRes(
            ConversationContracts.toMessage(message),
            ConversationContracts.toState(change.state),
        )
    }

    suspend fun setTyping(actor: SupportUserActor, message: SetTypingMsg) {
        publishChange(requireConversation().setTyping(actor.participantId, message.isTyping))
    }

    suspend fun close(
        actor: SupportUserActor,
        request: CloseConversationReq,
    ): CloseConversationRes {
        val change = requireConversation().close(actor.participantId, request.reason)
        publishChange(change)
        return CloseConversationRes(ConversationContracts.toState(change.state))
    }

    // --8<-- [start:doc-sc-assign]
    private suspend fun assignAgent() {
        val conversation = requireConversation()
        val assigned = assignment.assignForConversation(conversation.conversationId) ?: return
        notifications.publishAssignedToRoster(
            directory.get(assigned.rosterActorId).actor,
            conversation.snapshot(),
        )
        logger.info(
            "support conversation: assigned. conversation={}, roster={}",
            conversation.conversationId,
            assigned.rosterActorId,
        )
    }

    // --8<-- [end:doc-sc-assign]

    private fun joinAgent(agent: SupportUserActor): ConversationChange {
        val conversation = requireConversation()
        val change =
            conversation.joinAgent(
                agent.participantId,
                agent.displayName,
                System.currentTimeMillis(),
            )
        agent.joinConversation(conversation.conversationId)
        actors[agent.participantId] = directory.get(agent.participantId).actor
        return change
    }

    private suspend fun publishChange(change: ConversationChange) {
        notifications.publish(change.events, actors)
        for (event in change.events) {
            logger.info(
                "supportchat-conversation status={} conversation={}",
                event.state.status,
                event.state.conversationId,
            )
        }
        if (change.events.any { it.kind == ConversationEventKind.Closed }) {
            assignment.releaseConversation(requireConversation().conversationId)
        }
    }

    private fun requireConversation(): Conversation =
        conversation ?: error("Conversation has not been created.")

    private companion object {
        private val logger = LoggerFactory.getLogger(ConversationSpot::class.java)
    }
}
